### ISSUE TITLE
Can't get the media to load from db

### ISSUE THREAD

**Issue by jayathu** [jayathu/TwitterAppClient#2]
@codepathreview @codepath

This has been bugging me for the past couple of days. I think I am following all the 'rules' for saving the objects properly. I can get the basic tweet texts, profile pics and user information from the db but I can't get the media to load. At this point, I have almost given up and I am going in circles trying to spot the issue. Would appreciate extra pair of eyes taking a look at it! 

To give you pointers to the code that's giving me issues: here are a couple of places you may want to start looking:

TimelineActivity::StoreTweetsToLocalDatabase at Line #323.
ExtendedEntities::createIfExists at Line #36

Thanks for the help!

**Comment by jayathu** [comment]
@codepathreview @codepath
Just narrowing down my question here - I guess I am unclear on how to fetch a list of something from the database when they are a collection of another model class.

For example, ExtendedEntities has a list of Media 
Media has VideoInfo
VideoInfo has a list of Variants.

I have public empty constructors for every single model and I am making sure that I "save" the inner objects _before_ I assign it to its parent class and saving the parent class. 

But here's where I am lost:
If I want the list of media from the extended entities object, do I perform a select query on the media instead of simply returning the list<media> ? If so, how can I constrain it to a particular user or a tweet id? Or am i just complicating things?

**Comment by nidhi1608** [comment]
I think this is the same question you posted on the group. As I mentioned, try restructuring your models like this:
https://github.com/google/gson/blob/master/UserGuide.md#nested-classes-including-inner-classes

**Comment by rogerhu** [comment]
Jay,

Saving empty constructors helps with your code running correctly, but ActiveAndroid doesn't out of the box handle saving/retrieving nested data structures for you.  Usually you will have to build this structure up yourself (i.e. get the tweet, get the extended entity id, and then fetch the media.  Your Media object for instance should probably have an integer column of the extended entity ID so you can relate it back. 

But regardless, you have to fetch this data and build it back up yourself.  Does that make sense?

**Comment by jayathu** [comment]
Hi Roger,

Does that mean I have to run Select() query with a ID constraint on
extended_entities? I did think of this possibility but then I wonder what
would I do in situations where I have no id information? For example
Video_info has an array of variants but the variants dont have ids to query
them with.

Thanks,
Jayashree

**Comment by rogerhu** [comment]
Yeah you'd have to create an ID by creating an extended entry row and mapping.

I've been talking to Nathan and Nidhi about this issue you're having, and the real crux is that many students like yourself have been trying to use Gson and create classes that map directly  to the Twitter API response data.  When it comes to persisting this data, it isn't a good idea to try to use the same class structure to store it on desk.  You end up in a scenario with multiple tables with different relationships.  ActiveAndroid and most ORM's are not suited for this kind of stuff for trying to recreate all the Java objects needed to reflect. Beyond the simplest cases, it is usually on you to figure out those relationships and recreating them.

The simpler way is to create another class that denormalizes this data (flattens it) and stores all the data in a way that you can manage.   In other words, create model classes that manage the API response and 
one classes that reflect how they will actually be used in your app.  You then would need someway to translate between the two, but at least there is a separation of responsibilities at the expense of more overhead of two separate types of classes.

**Comment by nidhi1608** [comment]
Jayashree,

Just wanted to let you know that @rogerhu and I are looking into this issue. We'll let you know here once we find something. But as he mentioned, its probably best to flatten out the structure.

**Comment by nidhi1608** [comment]
It's not very clear how AA handles deep nested structures as Roger mentioned. You have a  few of options here:
1. As Roger mentioned, you may try manually adding unique IDs to all entities and also their foreign constraints. This will also require you to fetch the entities separately and build the object again yourself.
2. Ditch gson and have a flat structure with just one tweet class with all properties. This will greatly simplify your architecture.
3. Ditch AA and continue using gson if you don't want to refactor at this point.
4. Create static inner classes for your entities, like so - https://gist.github.com/nidhi1608/e4dd10f681b6cd623903

**Comment by jayathu** [comment]
Yup I realized I may have to ditch one of these, specially, given the
complex structuring of the tweet json objects. I think I might be better
off with #2 flattening my models - it'll be much more readable and
maintainable.

Thanks so much for looking into this! I truly appreciate it :)
